File: rainkey.py
# rainkey.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2025 Anonymous
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/agpl-3.0.html>.
#
# Notes: Integrates wise_transforms.py for mapping hex color grid based on spiral's hop path vectors and spectrum.
# Applies BitWise for binary pruning, HexWise for reversible rotations, HashWise for immutable entropy.
# Generates a color grid visualization where each hop vector influences RGB values via transformations.
import numpy as np
import matplotlib.pyplot as plt
import time
import random
import math
import argparse
import hashlib
import mpmath

# Set precision for mpmath
mpmath.mp.dps = 19

# wise_transforms functions
def bitwise_transform(data, bits=16):
    """BitWise: Raw binary ops (mask, NOT flip) for -1 pruning."""
    int_data = int.from_bytes(data.encode(), 'big') % (1 << bits)
    mask = (1 << bits) - 1
    mirrored = (~int_data) & mask  # Bitwise mirror
    return bin(mirrored)[2:].zfill(bits)  # Binary string

def hexwise_transform(data, angle=137.5):
    """HexWise: String/hex rotations/mirrors for 0 privacy (reversible)."""
    hex_data = data.encode().hex()
    mirrored = hex_data + hex_data[::-1]  # Palindromic mirror
    shift = int(angle % len(mirrored))
    rotated = mirrored[shift:] + mirrored[:shift]  # Rotate
    return rotated

def hashwise_transform(data):
    """HashWise: SHA1664 sponge perms for +1 culture (immutable entropy)."""
    base_hash = hashlib.sha512(data.encode()).digest()
    mp_state = mpmath.mpf(int(base_hash.hex(), 16))
    for _ in range(4):  # Perms (sqrt * PHI)
        mp_state = mpmath.sqrt(mp_state) * mpmath.phi
    partial = mpmath.nstr(mp_state, 1664 // 4)
    final_hash = hashlib.sha256(partial.encode()).hexdigest()
    entropy = int(mpmath.log(mp_state, 2))
    return final_hash, entropy

# Expanded QWERTY layout (4 rows, including numbers and letters)
qwerty = [
    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0'],
    ['Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'],
    ['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L', ';'],
    ['Z', 'X', 'C', 'V', 'B', 'N', 'M', ',', '.', '/']
]

# Strict hex mapping (0-9, a-f only, mapped to all keys)
hex_map = {
    '0': '0', '1': '1', '2': '2', '3': '3', '4': '4', '5': '5', '6': '6', '7': '7', '8': '8', '9': '9',
    'A': 'a', 'B': 'b', 'C': 'c', 'D': 'd', 'E': 'e', 'F': 'f', 'G': 'g', 'H': 'h', 'I': 'i', 'J': 'j',
    'K': 'k', 'L': 'l', 'M': 'm', 'N': 'n', 'O': 'o', 'P': 'p', 'Q': 'q', 'R': 'r', 'S': 's', 'T': 't',
    'U': 'u', 'V': 'v', 'W': 'w', 'X': 'x', 'Y': 'y', 'Z': 'z', ';': 's', ',': 'c', '.': 'd', '/': 'f'
}

# Key positions (row, col)
key_pos = {key: (r, c) for r, row in enumerate(qwerty) for c, key in enumerate(row)}

def generate_spiral_sequence(start_key, time_factor, num_hops=20, kappa=1.0):
    """Generate a forward spiral sequence, unique to the moment."""
    if start_key not in key_pos:
        raise ValueError(f"Invalid start key: {start_key}")
    
    r, c = key_pos[start_key]
    sequence = [start_key]
    theta = 0.0
    visited = {start_key}
    
    for hop in range(1, num_hops):
        theta += (137.5 * math.pi / 180) / (hop * kappa) + time_factor
        distance = hop / kappa
        dx = math.cos(theta) * distance
        dy = math.sin(theta) * distance
        new_r = int((r + dy) % len(qwerty))
        new_c = int((c + dx) % len(qwerty[0]))
        new_key = qwerty[new_r][new_c]
        
        if new_key not in visited and new_key != sequence[0]:
            sequence.append(new_key)
            visited.add(new_key)
            r, c = new_r, new_c
        elif len(visited) < len(key_pos):
            directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
            dr, dc = random.choice(directions)
            adj_r = (r + dr) % len(qwerty)
            adj_c = (c + dc) % len(qwerty[0])
            adj_key = qwerty[adj_r][adj_c]
            if adj_key not in visited and adj_key != sequence[0]:
                sequence.append(adj_key)
                visited.add(adj_key)
                r, c = adj_r, adj_c
        if len(sequence) >= num_hops:
            break
    
    # Pad to num_hops if not reached
    while len(sequence) < num_hops:
        available_keys = [k for k in key_pos if k not in visited and k != start_key]
        if available_keys:
            new_key = random.choice(available_keys)
            sequence.append(new_key)
            visited.add(new_key)
        else:
            sequence.append(random.choice(list(key_pos.keys())))
    
    return sequence

def pollard_kangaroo_on_grid(start_pos, target_pos, steps=800):
    """Enhanced Pollard's Kangaroo with increased steps and dynamic jumps including diagonals."""
    grid_rows, grid_cols = len(qwerty), len(qwerty[0])
    order_approx = grid_rows * grid_cols
    m = int(math.sqrt(order_approx)) + 15  # Increased m for better coverage
    # Jumps including diagonal moves
    jumps = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]
    tame_map = {}
    x_t, y_t = start_pos
    for i in range(m):
        pos = (x_t % grid_rows, y_t % grid_cols)
        if pos not in tame_map:
            tame_map[pos] = i
        j = hash(str(pos) + str(i)) % len(jumps)
        dr, dc = jumps[j]
        x_t = (x_t + dr) % grid_rows
        y_t = (y_t + dc) % grid_cols
    x_w, y_w = target_pos
    for i in range(m * 4):  # Increased search range
        pos = (x_w % grid_rows, y_w % grid_cols)
        if pos in tame_map:
            return tame_map[pos] + i
        j = hash(str(pos) + str(i)) % len(jumps)
        dr, dc = jumps[j]
        x_w = (x_w + dr) % grid_rows
        y_w = (y_w + dc) % grid_cols
    # Fallback to Manhattan distance if no path found
    return abs(start_pos[0] - target_pos[0]) + abs(start_pos[1] - target_pos[1])

def generate_spectrum_kappa(sequence):
    """Generate strict hex spectrum kappa (0-9, a-f), padded to 16 chars."""
    spectrum = ''.join(hex_map.get(k.upper(), '0') for k in sequence)
    spectrum = (spectrum + '0' * (16 - len(spectrum)))[:16]
    return f"0x{spectrum.lower()}"

def map_hex_color_grid(sequence, hop_distance, spectrum_kappa):
    """Map a hex color grid based on spiral hop path vectors using wise transformations."""
    grid_size = (len(qwerty), len(qwerty[0]))
    color_grid = np.zeros(grid_size + (3,))  # RGB grid
    
    # Apply wise transformations to spectrum_kappa
    bit_wise = bitwise_transform(spectrum_kappa)
    hex_wise = hexwise_transform(spectrum_kappa)
    hash_wise, _ = hashwise_transform(spectrum_kappa)
    
    for i in range(len(sequence) - 1):
        current = sequence[i]
        next_key = sequence[i + 1]
        current_pos = next((r, c) for r, row in enumerate(qwerty) for c, val in enumerate(row) if val == current)
        next_pos = next((r, c) for r, row in enumerate(qwerty) for c, val in enumerate(row) if val == next_key)
        
        # Hop vector (direction and magnitude)
        vector = (next_pos[0] - current_pos[0], next_pos[1] - current_pos[1])
        vector_str = f"{vector[0]},{vector[1]}"
        
        # Transform vector to color
        hash_color, _ = hashwise_transform(vector_str)
        r = int(hash_color[:2], 16) % 256 / 255  # Red from first two hex digits
        g = int(hash_color[2:4], 16) % 256 / 255  # Green from next two
        b = int(hash_color[4:6], 16) % 256 / 255  # Blue from next two
        color = (r, g, b)
        
        # Apply color to grid at current position
        color_grid[current_pos[0], current_pos[1]] = color
    
    return color_grid

def visualize_color_grid(color_grid):
    """Visualize the hex color grid."""
    fig, ax = plt.subplots(figsize=(10, 8))
    ax.imshow(color_grid)
    ax.set_title("Hex Color Grid based on Spiral Hop Path Vector Spectrum")
    ax.set_xticks(np.arange(-0.5, len(qwerty[0]), 1), minor=True)
    ax.set_yticks(np.arange(-0.5, len(qwerty), 1), minor=True)
    ax.grid(which="minor", color="w", linestyle='-', linewidth=1)
    plt.show()

def main():
    parser = argparse.ArgumentParser(description="Rainkey: QWERTY Hexwise Random Input Generator with Spiral Rainbow and Color Grid")
    parser.add_argument("--start-key", type=str, default="Q", help="Starting key on QWERTY")
    parser.add_argument("--num-hops", type=int, default=20, help="Number of hops in sequence")
    parser.add_argument("--kappa", type=float, default=1.0, help="Kappa for flattening")
    args = parser.parse_args()

    # Time factor unique to moment (forward only, >0)
    time_factor = (time.time() % 1) + 0.01

    # Generate sequence
    sequence = generate_spiral_sequence(args.start_key.upper(), time_factor, args.num_hops, args.kappa)
    print("Generated Sequence:", sequence)

    # Kangaroo hop distance
    start_pos = key_pos[sequence[0]]
    target_pos = key_pos[sequence[-1]]
    hop_distance = pollard_kangaroo_on_grid(start_pos, target_pos)
    print(f"Kangaroo hop distance to end key: {hop_distance}")

    # Spectrum kappa
    spectrum_kappa = generate_spectrum_kappa(sequence)
    print("Spectrum Kappa:", spectrum_kappa)

    # Map and visualize color grid
    color_grid = map_hex_color_grid(sequence, hop_distance, spectrum_kappa)
    visualize_color_grid(color_grid)

if __name__ == "__main__":
    main()

File: rainkey.c
// rainkey.c
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 Anonymous
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/agpl-3.0.html>.
//
// Notes: Generates a spiral sequence on a QWERTY grid, calculates kangaroo hop distance,
//        generates a spectrum kappa, and computes Shannon entropy. Adapted from rainkey.py
//        for C, without visualization or wise_transforms integration.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

// QWERTY layout (4x10 grid)
char qwerty[4][10] = {
    {'1', '2', '3', '4', '5', '6', '7', '8', '9', '0'},
    {'Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'},
    {'A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L', ';'},
    {'Z', 'X', 'C', 'V', 'B', 'N', 'M', ',', '.', '/'}
};

// Hex mapping (0-9, a-f for all keys)
char hex_map[256] = {0};
void init_hex_map() {
    char hex_chars[] = "0123456789abcdef";
    int i, j;
    for (i = 0; i < 10; i++) hex_map[qwerty[0][i]] = hex_chars[i];
    for (i = 0; i < 10; i++) hex_map[qwerty[1][i]] = hex_chars[i + 10];
    for (i = 0; i < 9; i++) hex_map[qwerty[2][i]] = hex_chars[i + 20];
    hex_map[qwerty[2][9]] = 's';  // ';'
    for (i = 0; i < 7; i++) hex_map[qwerty[3][i]] = hex_chars[i + 27];
    hex_map[qwerty[3][7]] = 'c';  // ','
    hex_map[qwerty[3][8]] = 'd';  // '.'
    hex_map[qwerty[3][9]] = 'f';  // '/'
}

int key_pos[4][10][2] = {
    {{0, 0}, {0, 1}, {0, 2}, {0, 3}, {0, 4}, {0, 5}, {0, 6}, {0, 7}, {0, 8}, {0, 9}},
    {{1, 0}, {1, 1}, {1, 2}, {1, 3}, {1, 4}, {1, 5}, {1, 6}, {1, 7}, {1, 8}, {1, 9}},
    {{2, 0}, {2, 1}, {2, 2}, {2, 3}, {2, 4}, {2, 5}, {2, 6}, {2, 7}, {2, 8}, {2, 9}},
    {{3, 0}, {3, 1}, {3, 2}, {3, 3}, {3, 4}, {3, 5}, {3, 6}, {3, 7}, {3, 8}, {3, 9}}
};

int find_position(char key, int *r, int *c) {
    for (int i = 0; i < 4; i++)
        for (int j = 0; j < 10; j++)
            if (qwerty[i][j] == key) {
                *r = i;
                *c = j;
                return 1;
            }
    return 0;
}

void generate_spiral_sequence(char start_key, int num_hops, float kappa, char *sequence) {
    if (num_hops > 20) num_hops = 20;  // Prevent buffer overflow
    int r, c;
    if (!find_position(start_key, &r, &c)) {
        strcpy(sequence, "Invalid start key");
        return;
    }
    sequence[0] = start_key;
    int visited[4][10] = {0};
    visited[r][c] = 1;
    int seq_len = 1;
    float theta = 0.0;
    srand(time(NULL) * clock());  // Unique seed per run
    float time_factor = (float)(rand() % 1000) / 1000.0 + 0.01;
    for (int hop = 1; hop < num_hops; hop++) {
        theta += (137.5 * M_PI / 180.0) / (hop * kappa) + time_factor;
        float distance = hop / kappa;
        float dx = cos(theta) * distance;
        float dy = sin(theta) * distance;
        int new_r = (int)((r + dy) + 0.5) % 4;
        int new_c = (int)((c + dx) + 0.5) % 10;
        char new_key = qwerty[new_r][new_c];
        if (!visited[new_r][new_c] && new_key != start_key) {
            sequence[seq_len++] = new_key;
            visited[new_r][new_c] = 1;
            r = new_r;
            c = new_c;
        } else if (seq_len < 40) {
            int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
            int dr = dirs[rand() % 4][0];
            int dc = dirs[rand() % 4][1];
            int adj_r = (r + dr + 4) % 4;
            int adj_c = (c + dc + 10) % 10;
            char adj_key = qwerty[adj_r][adj_c];
            if (!visited[adj_r][adj_c] && adj_key != start_key) {
                sequence[seq_len++] = adj_key;
                visited[adj_r][adj_c] = 1;
                r = adj_r;
                c = adj_c;
            }
        }
        if (seq_len >= num_hops) break;
    }
    while (seq_len < num_hops) {
        int r_new, c_new;
        do {
            r_new = rand() % 4;
            c_new = rand() % 10;
        } while (visited[r_new][c_new] || qwerty[r_new][c_new] == start_key);
        sequence[seq_len++] = qwerty[r_new][c_new];
        visited[r_new][c_new] = 1;
    }
    sequence[seq_len] = '\0';
}

int pollard_kangaroo_on_grid(int start_r, int start_c, int target_r, int target_c, int steps) {
    int grid_rows = 4, grid_cols = 10;
    int m = (int)sqrt(grid_rows * grid_cols) + 15;
    int jumps[8][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};
    int tame_map[4][10] = {-1};
    int x_t = start_r, y_t = start_c;
    for (int i = 0; i < m; i++) {
        int pos_r = x_t % grid_rows;
        int pos_c = y_t % grid_cols;
        if (tame_map[pos_r][pos_c] == -1) tame_map[pos_r][pos_c] = i;
        int j = rand() % 8;
        x_t = (x_t + jumps[j][0]) % grid_rows;
        y_t = (y_t + jumps[j][1]) % grid_cols;
    }
    int x_w = target_r, y_w = target_c;
    for (int i = 0; i < m * 4; i++) {
        int pos_r = x_w % grid_rows;
        int pos_c = y_w % grid_cols;
        if (tame_map[pos_r][pos_c] != -1) return tame_map[pos_r][pos_c] + i;
        int j = rand() % 8;
        x_w = (x_w + jumps[j][0]) % grid_rows;
        y_w = (y_w + jumps[j][1]) % grid_cols;
    }
    // Fallback to Manhattan distance
    return abs(start_r - target_r) + abs(start_c - target_c);
}

void generate_spectrum_kappa(char *sequence, char *kappa, int len) {
    int i = 0;
    for (; sequence[i] && i < len; i++) {
        kappa[i] = hex_map[(unsigned char)sequence[i]];
    }
    while (i < 16) kappa[i++] = '0';
    kappa[16] = '\0';
    for (i = 0; i < 16; i++) kappa[i] = tolower(kappa[i]);
    sprintf(kappa, "0x%s", kappa + (kappa[0] == '0' && kappa[1] == 'x' ? 2 : 0));
}

double calculate_shannon_entropy(char *sequence) {
    int freq[256] = {0};
    int total = strlen(sequence);
    for (int i = 0; i < total; i++) freq[(unsigned char)sequence[i]]++;
    double entropy = 0.0;
    for (int i = 0; i < 256; i++) {
        if (freq[i] > 0) {
            double p = (double)freq[i] / total;
            entropy -= p * log2(p);
        }
    }
    return entropy;
}

int main(int argc, char *argv[]) {
    char start_key = (argc > 1) ? argv[1][0] : 'Q';
    int num_hops = (argc > 2) ? atoi(argv[2]) : 20;
    if (num_hops > 20) num_hops = 20;  // Prevent buffer overflow
    float kappa = (argc > 3) ? atof(argv[3]) : 1.0;

    init_hex_map();
    char sequence[21];  // Max 20 chars + null
    generate_spiral_sequence(start_key, num_hops, kappa, sequence);
    printf("Generated Sequence: %s\n", sequence);

    int start_r, start_c, target_r, target_c;
    find_position(sequence[0], &start_r, &start_c);
    find_position(sequence[num_hops - 1], &target_r, &target_c);
    int hop_distance = pollard_kangaroo_on_grid(start_r, start_c, target_r, target_c, 800);
    printf("Kangaroo hop distance to end key: %d\n", hop_distance);

    char spectrum_kappa[18];  // 0x + 16 chars + null
    generate_spectrum_kappa(sequence, spectrum_kappa, num_hops);
    printf("Spectrum Kappa: %s\n", spectrum_kappa);

    double entropy = calculate_shannon_entropy(sequence);
    printf("Shannon Entropy: %.4f bits\n", entropy);

    return 0;
}
